[BUG] Fix RecursiveReductionForecaster Bug

Sohaib-Ahmed21 opened the pull request:
<!--
Welcome to sktime, and thanks for contributing!
Please have a look at our contribution guide:
https://www.sktime.net/en/latest/get_involved/contributing.html
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.

Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests.
If no issue exists, you can open one here: https://github.com/sktime/sktime/issues
-->
Fixes: #7852 


#### What does this implement/fix? Explain your changes.
<!--
A clear and concise description of what you have implemented.
-->
Removed imputer as not needed in logic of `RecursiveReductionForecaster`.


- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

Sohaib-Ahmed21 commented:
Errors seem unrelated to this PR, being fixed in https://github.com/sktime/sktime/pull/7846.

yarnabrina commented:
I haven't tested, but this imputation method change will fix the additional row in reduction that @ericjb mentioned on #7852 ?

ericjb commented:
I tested this fix with and without X_exog. It looks good to me. Thanks!

fkiraly reviewed (changes requested):
This does not look right, it removes the imputation method.

Sohaib-Ahmed21 commented:
> This does not look right, it removes the imputation method.

@ericjb @fkiraly can you briefly explain the necessity of imputer in `RecursiveReductionForecaster`. Imo forecaster will be working fine without it.

Sohaib-Ahmed21 commented:
Thanks for the review @ericjb.
> I believe that imputation, if used at all, should be applied to y and X before the reduction is made to tabular form for ML models.

Yes, same are my thoughts. @fkiraly can confirm.

> 2\. in the additional comments, that the bug was also present in the DirectReductionForecaster. This PR has not addressed that case.

Oh, I didn't look at it. Will address that soon.

Sohaib-Ahmed21 commented:
The core problem is that imputer makes **_illogical_** training samples at the beginning which contributes to wrong results.

Sohaib-Ahmed21 commented:
@fkiraly @ericjb I have tested `check_estimator` for `DirectReductionForecaster` and `RecursiveReductionForecaster` and all tests are passing with current changes. I made some changes for  logical training samples:

- Removed imputer as it was making illogical training samples at the beginning which contributes to wrong results.
- Imputer was only needed at one place in `ReducerTransform` in case of `X_treatment` being `concurrent` in `DirectReductionForecaster` and now it is properly added to that in `ReducerTransform`. Now it applies imputer in `ReducerTransform` to only exogeneous data made there from `YtoX()`.
- Make some changes for correct logic after removing imputer like _handling offsets properly in case of `windows_identical = True` and `X_treatment` being `concurrent` for in sample predictions_.

Any suggestions/reviews will be helpful. Thanks!

hliebert commented:
I'm not sure if removing the imputation altogether is the right way forward, but I agree that it isn't a sensible default. I was puzzled when I first realized the missing values created by lagging were silently backfilled.

hliebert commented:
Thanks for looking into this. A small ask: Since you are changing the offset calculation in `DirectReductionForecaster`, can you change l. 1991 to `return self._predict_multiple(X=X, fh=fh)` - i.e. use the actual `X` instead of the X seeen during training and verify whether out of predictions on new X work with your fixes. I've been meaning to do this.

fkiraly reviewed (changes requested):
Removing the imputation is worth discussing, but we should not do it without deprecation. See here:
https://www.sktime.net/en/stable/developer_guide/deprecation.html

No user code should break.
Even though the warning message is still there, people are already relying on it.

Other questions:

* can you explain how this solves the "extra row in `X`" issue?
* why are you changing the `Lag` transformer? This will break in the `None` case. If no tests detected that you broke the transformation, please add one in `get_test_params`.

fkiraly commented:
> Thanks for looking into this. A small ask: Since you are changing the offset calculation in `DirectReductionForecaster`, can you change l. 1991 to `return self._predict_multiple(X=X, fh=fh)` - i.e. use the actual `X` instead of the X seeen during training and verify whether out of predictions on new X work with your fixes. I've been meaning to do this.

that might be a good idea.

Sohaib-Ahmed21 commented:
> Thanks for looking into this. A small ask: Since you are changing the offset calculation in `DirectReductionForecaster`, can you change l. 1991 to `return self._predict_multiple(X=X, fh=fh)` - i.e. use the actual `X` instead of the X seeen during training and verify whether out of predictions on new X work with your fixes. I've been meaning to do this.

I have tested this and its working fine with current changes and `check_estimator` is also passing all the tests.

Sohaib-Ahmed21 commented:
Thanks for the review @fkiraly.
> Removing the imputation is worth discussing, but we should not do it without deprecation. See here: https://www.sktime.net/en/stable/developer_guide/deprecation.html  
>   
I'll do it after we're on the same page.  

> Other questions:  
>   
> * can you explain how this solves the "extra row in `X`" issue?  

The issue arises because the imputer generates **artificial** training samples at the beginning, leading to incorrect results. By removing the imputer, we prevent these **illogical** training samples from being created in the first place.  

> * why are you changing the `Lag` transformer?  

The imputer was only required in `ReducerTransform` when `X_treatment` was **concurrent** in `DirectReductionForecaster`. Instead of applying imputation at the `Lag` transformer level, it is now **properly** handled within `ReducerTransform`, ensuring it is only applied to exogenous data derived from `YtoX()`.  

> * This will break in the `None` case.  

Previously, all calls to `ReducerTransform` used the default `impute_method` (`"bfill"`), so nothing was affected. However, I have now explicitly handled cases where `impute_method` is `None` to prevent any unexpected behavior.  

> * If no tests detected that you broke the transformation, please add one in `get_test_params`.  

I have updated the logic with an `if-else` condition to handle cases where `impute_method` is `None`. Additionally, I have added tests to `get_test_params` to catch any similar breaking changes in the future.

Sohaib-Ahmed21 commented:
Closing as completed and confirmed via https://github.com/sktime/sktime/issues/7852#issuecomment-2708346467
